### PR TITLE
Feature/component specific lists

### LINKED ISSUE #7: Create component specific lists of entities
This is what entityplus is doing to get big performance boosts. The idea is you have the full lists of entities, and then for each component you keep track of entities with that component. When you go to search for entities with component X, you look at the list of entities with component X first, which likely is less than the total list of entities, and thus less to search through, and thus faster to search.

This becomes huge when you have rare components (like AI, or items) that only a dozen entities in the ECS of tens of thousands of entities.

### PR DESCRIPTION
**Pull request by misnor**
Closes #7 

Adds component specific lists that greatly reduce the search time for entities with rare components added to them.

Example:

- 1000 Entities
  - 900 with Identity
  - 100 with AI

Using `EntitiesWith(Identity(), AI());` we do NOT search the entire entity list anymore (1,000 entries). We detect that the smallest component being requested is `AI` and start there - narrowing the search down to 100.

Some other various modifications and improvements on performance were made as well:

- Component Names via recursion in EntitiesWith is switched to pass a pointer to the vector of names instead of reference
- Checking the map if a component is registered at all is moved off an iterative approach to a quick 'count' check.
- Benchmark has been reworked to actually be correct now

### CODE REVIEW

File: src/Manager.hpp [hunk 1]
```diff
@@ -155,6 +150,19 @@ inline void ECS::AddComponent(Entity entity, T component)
...
+			if (iter != this->individualComponentVecs.end())
+			{
+				iter->second.push_back(e);
+			}
+			else
+			{
+				std::vector<Entity> entityList;
+				entityList.push_back(e);
+				this->individualComponentVecs.insert(std::pair<std::string, std::vector<Entity>>(componentName, entityList));
+				//this->individualComponentVecs.insert(std::make_pair(componentName, e));
```
Comment (jddeal) [inline src/Manager.hpp]: nukeable line?

File: src/Manager.hpp [hunk 2]
```diff
@@ -219,14 +227,29 @@ inline std::vector<Entity*> ECS::EntitiesWith(Ts&& ...types)
...
+	
+	if (componentNames.size() == 0)
+	{
+		// Asking for all entities
+		std::vector<Entity*> requestedEntities;
+		for (auto e : this->entities)
+		{
+			requestedEntities.push_back(&e);
+		}
+		return requestedEntities;
```
Comment (jddeal) [inline src/Manager.hpp]: Minor thing on this block above, but given that we're just sending back the entire list of entities, we can probably just do `return this->entities`

Comment (misnor) [reply src/Manager.hpp]: return turn is of type `std::vector<Entity*>`

`entities` is of type `std::vector<Entity>` - is there a more obvious way to convert it the expected (pointer) type?

Comment (misnor) [reply src/Manager.hpp]: Return types are not the same, resolving comment per conversation.

**Pull request merged**